File: arrays/merge-intervals/merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
function merge(intervals) {
  const ans = [];

  intervals.sort((a, b) => a[0] - b[0]);

  for (let i = 0; i < intervals.length; i++) {
    const starts = intervals[i][0];
    const ends = intervals[i][1];

    if (!ans.length || ans[ans.length - 1][1] < starts) {
      ans.push(intervals[i]);
    } else {
      const newEndingInterval = Math.max(ans[ans.length - 1][1], ends);
      ans[ans.length - 1][i] = newEndingInterval;
    }
  }
  return ans;
}

File: arrays/merge-intervals/merge-intervals.java
class Solution {
  public int[][] merge(int[][] intervals) {
    List<int[]> ans = new ArrayList<>();

    Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

    for (int[] interval : intervals) {
      int starts = interval[0];
      int ends = interval[1];

      if (ans.isEmpty() || ans.get(ans.size() - 1)[1] < starts) {
        ans.add(interval);
      } else {
        int newEndingTime = Math.max(ans.get(ans.size() - 1)[1], ends);
        ans.get(ans.size() - 1)[1] = newEndingTime;
      }
    }

    return ans.toArray(new int[ans.size()][2]);
  }
}
